File: src/parsers/factory.py
"""
Parser factory: choose parser by file extension and return extracted text.
"""

from pathlib import Path

from . import pdf_parser
from . import docx_parser
from . import txt_parser
from . import xlsx_parser

_EXT_MAP = {
    ".pdf": pdf_parser.parse_pdf,
    ".docx": docx_parser.parse_docx,
    ".txt": txt_parser.parse_txt,
    ".xlsx": xlsx_parser.parse_xlsx,
}


def parse_file(file_path: str) -> str:
    """
    Select parser by file extension and return extracted text.
    Raises ValueError if extension is not supported.
    """
    ext = Path(file_path).suffix.lower()
    parser_fn = _EXT_MAP.get(ext)
    if parser_fn is None:
        raise ValueError(f"Unsupported file type: {ext} (path: {file_path})")
    return parser_fn(file_path)

File: src/parsers/factory.cpp
#include "document_ingestion/parsers.hpp"
#include "document_ingestion/parsers_internal.hpp"
#include <filesystem>
#include <stdexcept>
#include <map>

namespace document_ingestion {

std::string parse_file(const std::string& file_path) {
  std::string ext = std::filesystem::path(file_path).extension().string();
  for (auto& c : ext) c = static_cast<char>(std::tolower(static_cast<unsigned char>(c)));

  if (ext == ".txt") return parse_txt(file_path);
  if (ext == ".pdf") return parse_pdf(file_path);
  if (ext == ".docx") return parse_docx(file_path);
  if (ext == ".xlsx") return parse_xlsx(file_path);

  throw std::runtime_error("Unsupported file type: " + ext + " (path: " + file_path + ")");
}

}  // namespace document_ingestion
